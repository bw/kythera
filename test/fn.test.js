require("jest")
const common = require("./common")

let i = 0

afterEach(() => {
	i += 1
})

describe("Function examples", () => {
	let program = common.parseFile("test/vars.ky")
	i = 0

	let t = (name) => {
		test(name, () => expect(program[i]).toMatchSnapshot())
	}

	// this could probably stand more specific testing
	t("Anonymous function and call")

	t("Fibonacci")
	t("Fibonacci (call)")

	t("Fizzbuzz and call")
})