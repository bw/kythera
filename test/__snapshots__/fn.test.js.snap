// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Functions [COMPL] Function with no parameters (with space) 1`] = `"let nullFn = new KYTHERA.value(() => {return new KYTHERA.value(null, KYTHERA.type.PRIMITIVES[\\"null\\"]);}, new KYTHERA.type(\\"fn\\", { parameters: [], returns: KYTHERA.type.PRIMITIVES[\\"null\\"]}));"`;

exports[`Functions [COMPL] Function with no parameters 1`] = `"let nullFn = new KYTHERA.value(() => {return new KYTHERA.value(null, KYTHERA.type.PRIMITIVES[\\"null\\"]);}, new KYTHERA.type(\\"fn\\", { parameters: [], returns: KYTHERA.type.PRIMITIVES[\\"null\\"]}));"`;

exports[`Functions [EVALT] Function with no parameters (with space) 1`] = `
Object {
  "nullFn":  {
    "type":  {
      "baseType": "fn",
      "structure": Object {
        "parameters": Array [],
        "returns":  {
          "baseType": "null",
        },
      },
    },
    "value": [Function],
  },
}
`;

exports[`Functions [EVALT] Function with no parameters 1`] = `
Object {
  "nullFn":  {
    "type":  {
      "baseType": "fn",
      "structure": Object {
        "parameters": Array [],
        "returns":  {
          "baseType": "null",
        },
      },
    },
    "value": [Function],
  },
}
`;

exports[`Functions [PARSE] Anonymous function and call 1`] = `
Array [
  ParseNode {
    "arguments": Array [
      ParseNode {
        "kind": "literal",
        "type": ParseNode {
          "baseType": "int",
          "kind": "type",
          "origin": "builtin",
        },
        "value": 10,
      },
    ],
    "kind": "call",
    "target": ParseNode {
      "body": Array [
        ParseNode {
          "identifier": "q",
          "kind": "let",
          "value": ParseNode {
            "kind": "literal",
            "type": ParseNode {
              "baseType": "int",
              "kind": "type",
              "origin": "builtin",
            },
            "value": 99,
          },
        },
        ParseNode {
          "identifier": "r",
          "kind": "let",
          "value": ParseNode {
            "kind": "identifier",
            "name": "q",
          },
        },
        ParseNode {
          "kind": "return",
          "value": ParseNode {
            "kind": "identifier",
            "name": "q",
          },
        },
      ],
      "kind": "literal",
      "parameters": Array [
        Object {
          "name": "a",
          "type": ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
        },
      ],
      "returns": ParseNode {
        "baseType": "int",
        "kind": "type",
        "origin": "builtin",
      },
      "type": ParseNode {
        "baseType": "fn",
        "kind": "type",
        "origin": "builtin",
        "parameters": Array [
          ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
        ],
        "returns": ParseNode {
          "baseType": "int",
          "kind": "type",
          "origin": "builtin",
        },
      },
    },
  },
]
`;

exports[`Functions [PARSE] Fibonacci 1`] = `
Array [
  ParseNode {
    "identifier": "fibo",
    "kind": "let",
    "value": ParseNode {
      "body": Array [
        ParseNode {
          "body": Array [
            ParseNode {
              "kind": "return",
              "value": ParseNode {
                "kind": "literal",
                "type": ParseNode {
                  "baseType": "int",
                  "kind": "type",
                  "origin": "builtin",
                },
                "value": 1,
              },
            },
          ],
          "condition": ParseNode {
            "kind": "binary",
            "left": ParseNode {
              "kind": "identifier",
              "name": "x",
            },
            "operator": "==",
            "right": ParseNode {
              "kind": "literal",
              "type": ParseNode {
                "baseType": "int",
                "kind": "type",
                "origin": "builtin",
              },
              "value": 1,
            },
          },
          "kind": "if",
        },
        ParseNode {
          "body": Array [
            ParseNode {
              "kind": "return",
              "value": ParseNode {
                "kind": "literal",
                "type": ParseNode {
                  "baseType": "int",
                  "kind": "type",
                  "origin": "builtin",
                },
                "value": 0,
              },
            },
          ],
          "condition": ParseNode {
            "kind": "binary",
            "left": ParseNode {
              "kind": "identifier",
              "name": "x",
            },
            "operator": "==",
            "right": ParseNode {
              "kind": "literal",
              "type": ParseNode {
                "baseType": "int",
                "kind": "type",
                "origin": "builtin",
              },
              "value": 0,
            },
          },
          "kind": "if",
        },
        ParseNode {
          "kind": "return",
          "value": ParseNode {
            "kind": "binary",
            "left": ParseNode {
              "arguments": Array [
                ParseNode {
                  "kind": "binary",
                  "left": ParseNode {
                    "kind": "identifier",
                    "name": "x",
                  },
                  "operator": "-",
                  "right": ParseNode {
                    "kind": "literal",
                    "type": ParseNode {
                      "baseType": "int",
                      "kind": "type",
                      "origin": "builtin",
                    },
                    "value": 1,
                  },
                },
              ],
              "kind": "call",
              "target": ParseNode {
                "kind": "identifier",
                "name": "fibo",
              },
            },
            "operator": "+",
            "right": ParseNode {
              "arguments": Array [
                ParseNode {
                  "kind": "binary",
                  "left": ParseNode {
                    "kind": "identifier",
                    "name": "x",
                  },
                  "operator": "-",
                  "right": ParseNode {
                    "kind": "literal",
                    "type": ParseNode {
                      "baseType": "int",
                      "kind": "type",
                      "origin": "builtin",
                    },
                    "value": 2,
                  },
                },
              ],
              "kind": "call",
              "target": ParseNode {
                "kind": "identifier",
                "name": "fibo",
              },
            },
          },
        },
      ],
      "kind": "literal",
      "parameters": Array [
        Object {
          "name": "x",
          "type": ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
        },
      ],
      "returns": ParseNode {
        "baseType": "int",
        "kind": "type",
        "origin": "builtin",
      },
      "type": ParseNode {
        "baseType": "fn",
        "kind": "type",
        "origin": "builtin",
        "parameters": Array [
          ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
        ],
        "returns": ParseNode {
          "baseType": "int",
          "kind": "type",
          "origin": "builtin",
        },
      },
    },
  },
  ParseNode {
    "arguments": Array [
      ParseNode {
        "kind": "literal",
        "type": ParseNode {
          "baseType": "int",
          "kind": "type",
          "origin": "builtin",
        },
        "value": 10,
      },
    ],
    "kind": "call",
    "target": ParseNode {
      "kind": "identifier",
      "name": "fibo",
    },
  },
]
`;

exports[`Functions [PARSE] Fizzbuzz and call 1`] = `
Array [
  ParseNode {
    "arguments": Array [
      ParseNode {
        "kind": "literal",
        "type": ParseNode {
          "baseType": "int",
          "kind": "type",
          "origin": "builtin",
        },
        "value": 20,
      },
    ],
    "kind": "call",
    "target": ParseNode {
      "body": Array [
        ParseNode {
          "identifier": "x",
          "kind": "let",
          "value": ParseNode {
            "kind": "literal",
            "type": ParseNode {
              "baseType": "int",
              "kind": "type",
              "origin": "builtin",
            },
            "value": 0,
          },
        },
        ParseNode {
          "body": Array [
            ParseNode {
              "body": Array [
                ParseNode {
                  "kind": "literal",
                  "type": ParseNode {
                    "baseType": "str",
                    "kind": "type",
                    "origin": "builtin",
                  },
                  "value": "fizzbuzz",
                },
              ],
              "condition": ParseNode {
                "kind": "binary",
                "left": ParseNode {
                  "kind": "binary",
                  "left": ParseNode {
                    "kind": "binary",
                    "left": ParseNode {
                      "kind": "identifier",
                      "name": "x",
                    },
                    "operator": "%",
                    "right": ParseNode {
                      "kind": "literal",
                      "type": ParseNode {
                        "baseType": "int",
                        "kind": "type",
                        "origin": "builtin",
                      },
                      "value": 3,
                    },
                  },
                  "operator": "==",
                  "right": ParseNode {
                    "kind": "literal",
                    "type": ParseNode {
                      "baseType": "int",
                      "kind": "type",
                      "origin": "builtin",
                    },
                    "value": 0,
                  },
                },
                "operator": "&&",
                "right": ParseNode {
                  "kind": "binary",
                  "left": ParseNode {
                    "kind": "binary",
                    "left": ParseNode {
                      "kind": "identifier",
                      "name": "x",
                    },
                    "operator": "%",
                    "right": ParseNode {
                      "kind": "literal",
                      "type": ParseNode {
                        "baseType": "int",
                        "kind": "type",
                        "origin": "builtin",
                      },
                      "value": 5,
                    },
                  },
                  "operator": "==",
                  "right": ParseNode {
                    "kind": "literal",
                    "type": ParseNode {
                      "baseType": "int",
                      "kind": "type",
                      "origin": "builtin",
                    },
                    "value": 0,
                  },
                },
              },
              "else": Array [
                ParseNode {
                  "body": Array [
                    ParseNode {
                      "kind": "literal",
                      "type": ParseNode {
                        "baseType": "str",
                        "kind": "type",
                        "origin": "builtin",
                      },
                      "value": "buzz",
                    },
                  ],
                  "condition": ParseNode {
                    "kind": "binary",
                    "left": ParseNode {
                      "kind": "binary",
                      "left": ParseNode {
                        "kind": "identifier",
                        "name": "x",
                      },
                      "operator": "%",
                      "right": ParseNode {
                        "kind": "literal",
                        "type": ParseNode {
                          "baseType": "int",
                          "kind": "type",
                          "origin": "builtin",
                        },
                        "value": 5,
                      },
                    },
                    "operator": "==",
                    "right": ParseNode {
                      "kind": "literal",
                      "type": ParseNode {
                        "baseType": "int",
                        "kind": "type",
                        "origin": "builtin",
                      },
                      "value": 0,
                    },
                  },
                  "else": Array [
                    ParseNode {
                      "body": Array [
                        ParseNode {
                          "kind": "literal",
                          "type": ParseNode {
                            "baseType": "str",
                            "kind": "type",
                            "origin": "builtin",
                          },
                          "value": "fizz",
                        },
                      ],
                      "condition": ParseNode {
                        "kind": "binary",
                        "left": ParseNode {
                          "kind": "binary",
                          "left": ParseNode {
                            "kind": "identifier",
                            "name": "x",
                          },
                          "operator": "%",
                          "right": ParseNode {
                            "kind": "literal",
                            "type": ParseNode {
                              "baseType": "int",
                              "kind": "type",
                              "origin": "builtin",
                            },
                            "value": 3,
                          },
                        },
                        "operator": "==",
                        "right": ParseNode {
                          "kind": "literal",
                          "type": ParseNode {
                            "baseType": "int",
                            "kind": "type",
                            "origin": "builtin",
                          },
                          "value": 0,
                        },
                      },
                      "else": Array [
                        ParseNode {
                          "kind": "identifier",
                          "name": "x",
                        },
                      ],
                      "kind": "if",
                    },
                  ],
                  "kind": "if",
                },
              ],
              "kind": "if",
            },
            ParseNode {
              "kind": "assign",
              "left": ParseNode {
                "kind": "identifier",
                "name": "x",
              },
              "operator": "=",
              "right": ParseNode {
                "kind": "binary",
                "left": ParseNode {
                  "kind": "identifier",
                  "name": "x",
                },
                "operator": "+",
                "right": ParseNode {
                  "kind": "literal",
                  "type": ParseNode {
                    "baseType": "int",
                    "kind": "type",
                    "origin": "builtin",
                  },
                  "value": 1,
                },
              },
            },
          ],
          "condition": ParseNode {
            "kind": "binary",
            "left": ParseNode {
              "kind": "identifier",
              "name": "x",
            },
            "operator": "<",
            "right": ParseNode {
              "kind": "identifier",
              "name": "count",
            },
          },
          "kind": "while",
        },
        ParseNode {
          "kind": "return",
          "value": ParseNode {
            "kind": "literal",
            "type": ParseNode {
              "baseType": "null",
              "kind": "type",
              "origin": "builtin",
            },
            "value": null,
          },
        },
      ],
      "kind": "literal",
      "parameters": Array [
        Object {
          "name": "count",
          "type": ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
        },
      ],
      "returns": ParseNode {
        "baseType": "null",
        "kind": "type",
        "origin": "builtin",
      },
      "type": ParseNode {
        "baseType": "fn",
        "kind": "type",
        "origin": "builtin",
        "parameters": Array [
          ParseNode {
            "baseType": "int",
            "kind": "type",
            "origin": "builtin",
          },
        ],
        "returns": ParseNode {
          "baseType": "null",
          "kind": "type",
          "origin": "builtin",
        },
      },
    },
  },
]
`;

exports[`Functions [PARSE] Function with no parameters (with space) 1`] = `
Array [
  ParseNode {
    "identifier": "nullFn",
    "kind": "let",
    "value": ParseNode {
      "body": Array [
        ParseNode {
          "kind": "return",
          "value": ParseNode {
            "kind": "literal",
            "type": ParseNode {
              "baseType": "null",
              "kind": "type",
              "origin": "builtin",
            },
            "value": null,
          },
        },
      ],
      "kind": "literal",
      "parameters": Array [],
      "returns": ParseNode {
        "baseType": "null",
        "kind": "type",
        "origin": "builtin",
      },
      "type": ParseNode {
        "baseType": "fn",
        "kind": "type",
        "origin": "builtin",
        "parameters": Array [],
        "returns": ParseNode {
          "baseType": "null",
          "kind": "type",
          "origin": "builtin",
        },
      },
    },
  },
]
`;

exports[`Functions [PARSE] Function with no parameters 1`] = `
Array [
  ParseNode {
    "identifier": "nullFn",
    "kind": "let",
    "value": ParseNode {
      "body": Array [
        ParseNode {
          "kind": "return",
          "value": ParseNode {
            "kind": "literal",
            "type": ParseNode {
              "baseType": "null",
              "kind": "type",
              "origin": "builtin",
            },
            "value": null,
          },
        },
      ],
      "kind": "literal",
      "parameters": Array [],
      "returns": ParseNode {
        "baseType": "null",
        "kind": "type",
        "origin": "builtin",
      },
      "type": ParseNode {
        "baseType": "fn",
        "kind": "type",
        "origin": "builtin",
        "parameters": Array [],
        "returns": ParseNode {
          "baseType": "null",
          "kind": "type",
          "origin": "builtin",
        },
      },
    },
  },
]
`;
